crates/ra_tt/src/lib.rs

```diff
@@ -56,21 +56,47 @@ pub struct Subtree {
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
 pub enum Delimiter {
-    Parenthesis,
-    Brace,
-    Bracket,
+    Parenthesis(TokenId),
+    Brace(TokenId),
+    Bracket(TokenId),
     None,
 }
 
+impl Delimiter {
+    pub fn id(&self) -> Option<TokenId> {
+        match self {
+            Delimiter::Parenthesis(id) => Some(*id),
+            Delimiter::Brace(id) => Some(*id),
+            Delimiter::Bracket(id) => Some(*id),
+            Delimiter::None => None,
+        }
+    }
+}
+
+impl Delimiter {
+    // Compare two delimiters but ignore token id
+    pub fn is_equal_delim(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Delimiter::Parenthesis(..), Delimiter::Parenthesis(..))
+            | (Delimiter::Brace(..), Delimiter::Brace(..))
+            | (Delimiter::Bracket(..), Delimiter::Bracket(..)) => true,
+            (Delimiter::None, Delimiter::None) => true,
+            _ => false,
+        }
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct Literal {
     pub text: SmolStr,
+    pub id: TokenId,
```

Comment:
Interesting!

I feel like it might make sense to pull `id`  up to the `Literal`, as every literal kind has an ID. OTOH, with this approach it's nice that leaves are self-contained types. I think we should stick with the current repr, but keep the other in mind as a plausible alternative